README.md
# canonicalwebteam.yaml-responses

[![Tests](https://github.com/canonical-web-and-design/canonicalwebteam.yaml-response/workflows/Tests/badge.svg)](https://github.com/canonical-web-and-design/canonicalwebteam.yaml-response/actions?query=workflow%3ATests)
[![Code coverage](https://codecov.io/gh/canonical-web-and-design/canonicalwebteam.yaml-responses/branch/master/graph/badge.svg)][codecov]

Easily serve `302` and `410` responses with simple YAML files.

This module contains Django and Flask helpers to serve ["302 Found"](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#301) redirect and ["410 Gone"](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#410) deleted responses from lists read from `redirects.yaml` and `deleted.yaml` files contained in the app directory.

## File format

**redirects.yaml**

``` yaml
hello: /world  # A simple redirect
example-(?P<name>.*): http://example.com/{name}  # A redirect with a regex replacement
```

**deleted.yaml**

``` yaml
deleted:
deleted/.*/regex:  # This will match "/deleted/{anything}/regex"
deleted/with/message:
  message: "Gone, gone, gone"  # This context data is passed directly to the template
```

## Usage

You can use this library with either Django or Flask.

### Django

Install the packege for Django as follows:

``` bash
pip install canonicalwebteam.yaml_responses[django]  # For Django helpers
```

Here's how to add URL patterns for redirects and deleted endpoints to Django:

#### `create_redirect_views` and `create_deleted_views` basic usage

This will read `redirects.yaml` and `deleted.yaml` pages in the project root.

For the "deleted" responses, it will look for a template called `410.html` in the default `templates/` directory.

``` python
# urls.py
from django.conf.urls import url
from canonicalwebteam.yaml_responses.django_helpers import (
    create_redirect_views,
    create_deleted_views
)

urlpatterns = create_redirect_views()  # Read redirects.yaml
urlpatterns += create_deleted_views()  # Read deleted.yaml

urlpatterns += ...  # The rest of your views
```

#### Options for `create_redirect_views`

- `path`: The path to the YAML file
- `permanent`: Return ["301 Moved Permanently"](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#301) statuses instead of 302

E.g.:

``` python
urlpatterns = create_redirect_views(
    path="config/permanent-redirects.yaml",
    permanent=True
)
urlpatterns += ...  # The rest of the views
```

#### Options for `create_deleted_views`

- `path`: The path to the YAML file
- `view_callback`: An alternative function to process Deleted responses

E.g.:

``` python
def deleted_callback(request, url_mapping, settings, *args, **kwargs):
    return render(request, "errors/410.html", url_mapping, status=410)

urlpatterns = create_deleted_views(
    path="config/deleted-paths.yaml",
    view_callback=deleted_callback
)
```

### Flask

Install the package for Flask as follows:

``` bash
pip install canonicalwebteam.yaml_responses[flask]  # For Flask helpers
```

Here's how to process redirects and deleted before other views are processed in Flask:

#### `prepare_redirects` and `prepare_deleted` basic usage

This will read `redirects.yaml` and `deleted.yaml` pages in the project root.

For the "deleted" responses, it will look for a template called `410.html` in the default `templates/` directory.

``` python
# app.py
from flask import Flask
from canonicalwebteam.yaml_responses.flask_helpers import (
    prepare_deleted,
    prepare_redirects,
)

app = Flask()

app.before_request(prepare_redirects())  # Read redirects.yaml
app.before_request(prepare_deleted())  # Read deleted.yaml
```

#### Options for `prepare_redirects`

- `path`: The path to the YAML file
- `permanent`: Return ["301 Moved Permanently"](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#301) statuses instead of 302

E.g.:

``` python
app.before_request(
    prepare_redirects(
        path=f"{parent_dir}/redirects.yaml",
        permanent=True
    )
)
```

#### Options for `prepare_deleted`

- `path`: The path to the YAML file
- `view_callback`: An alternative function to process Deleted responses

E.g.:

``` python
def deleted_callback(context):
    return render_template("errors/410.html"), 410

app.before_request(
    path=prepare_deleted(path="config/deleted-paths.yaml",
    view_callback=deleted_callback)
)
```

## Notes

This package has evolved from, and is intended to replace, the following projects:

- [canonicalwebteam.yaml-redirects](https://github.com/canonical-web-and-design/yaml-redirects)
- [canonicalwebteam.yaml-deleted-paths](https://github.com/canonical-web-and-design/yaml-deleted-paths)
- [canonicalwebteam.views-from-yaml](https://github.com/canonical-web-and-design/views-from-yaml)


[circleci]: https://circleci.com/gh/canonical-web-and-design/canonicalwebteam.yaml-responses "CircleCI build status"
[codecov]: https://codecov.io/gh/canonical-web-and-design/canonicalwebteam.yaml-responses "Code coverage"


CHANGES.txt
v1.0.0, 2018-09-21 -- Initial release: A merge of yaml-redirects, yaml-deleted-paths and views-from-yaml
v1.0.1, 2018-10-05 -- Support empty YAML files
v1.0.2, 2018-10-08 -- Handle extra arguments from views in Django without erroring
v1.1.0, 2018-10-09 -- Rename django->django_helpers flask->flask_helpers modules to avoid namespace clashes
v1.1.1, 2019-03-30 -- Security: Always specify loaders when using yaml.load()
v1.2.0, 2021-05-27 -- Improve query parameter support


tests/test_flask.py
# Core
import os
import unittest

# Local
from canonicalwebteam.yaml_responses.flask_helpers import (
    prepare_deleted,
    prepare_redirects,
)
from tests.fixtures.flask.app import (
    app_redirects,
    app_permanent_redirects,
    app_empty_redirects,
    app_empty_deleted,
    app_deleted,
    app_deleted_callback,
)


this_dir = os.path.dirname(os.path.realpath(__file__))


class TestFlaskRedirects(unittest.TestCase):
    def setUp(self):
        """
        Set up Flask app for testing
        """

        self.app_redirects = app_redirects.test_client()
        self.app_permanent_redirects = app_permanent_redirects.test_client()
        self.app_empty_redirects = app_empty_redirects.test_client()

    def test_missing_file(self):
        """
        When given a non-existent file path,
        prepare_redirects should not error
        """

        prepare_redirects(path="/tmp/non-existent-file.yaml")

    def test_not_found(self):
        """
        When Flask has redirects, check 404s still work
        """

        redirect_missing = self.app_redirects.get("/hello-missing")

        self.assertEqual(redirect_missing.status_code, 404)

    def test_not_found_empty_yaml(self):
        """
        When Flask is given an empty redirects file,
        check 404s still work
        """

        redirect_missing = self.app_empty_redirects.get("/hello-missing")

        self.assertEqual(redirect_missing.status_code, 404)

    def test_basic_redirect(self):
        """
        When Flask has redirects, check basic redirect works
        """

        redirect = self.app_redirects.get("/hello")

        self.assertEqual(redirect.status_code, 302)
        self.assertEqual(
            redirect.headers.get("Location"), "http://localhost/world"
        )

    def test_query_redirect(self):
        """
        When Flask has redirects, check redirect works with query string
        """

        redirect = self.app_redirects.get("/hello?name=world")
        redirect_query = self.app_redirects.get("/hello-query?name=world")

        self.assertEqual(redirect.status_code, 302)
        self.assertEqual(redirect_query.status_code, 302)
        self.assertEqual(
            redirect.headers.get("Location"),
            "http://localhost/world?name=world",
        )
        self.assertEqual(
            redirect_query.headers.get("Location"),
            "http://localhost/world?query=query&name=world",
        )

    def test_permanent_redirect(self):
        """
        When Flask has redirects, check permanent redirects work
        """

        redirect = self.app_permanent_redirects.get("/hello")

        self.assertEqual(redirect.status_code, 301)
        self.assertEqual(
            redirect.headers.get("Location"), "http://localhost/world"
        )

    def test_regex_redirect(self):
        """
        When Flask has redirects from redirects.yaml,
        check RegEx redirects work
        """

        robin_redirect = self.app_redirects.get("/example-robin")
        peter_redirect = self.app_redirects.get("/example-peter")

        self.assertEqual(robin_redirect.status_code, 302)
        self.assertEqual(
            robin_redirect.headers.get("Location"), "http://example.com/robin"
        )
        self.assertEqual(peter_redirect.status_code, 302)
        self.assertEqual(
            peter_redirect.headers.get("Location"), "http://example.com/peter"
        )

    def test_homepage_view(self):
        """
        When Flask has redirects from redirects.yaml
        check normal homepage view still works
        """

        homepage = self.app_redirects.get("/homepage")

        self.assertEqual(homepage.data, b"hello world")


class TestFlaskDeleted(unittest.TestCase):
    def setUp(self):
        """
        Set up Flask app for testing
        """

        self.app_deleted = app_deleted.test_client()
        self.app_deleted_callback = app_deleted_callback.test_client()
        self.app_empty_deleted = app_empty_deleted.test_client()

    def test_missing_file(self):
        """
        When given a non-existent file path,
        prepare_redirects should not error
        """

        prepare_deleted(path="/tmp/non-existent-file.yaml")

    def test_empty_file(self):
        """
        When given an empty file path,
        prepare_redirects should not error
        """

        prepare_deleted(path="{this_dir}/fixtures/empty.yaml")

    def test_not_found(self):
        """
        When Flask has deleteds, check 404s still work
        """

        deleted_missing = self.app_deleted.get("/deleted/missing")

        self.assertEqual(deleted_missing.status_code, 404)

    def test_not_found_empty_yaml(self):
        """
        When Flask is given an empty deleted file,
        check 404s still work
        """

        deleted_missing = self.app_empty_deleted.get("/deleted/missing")

        self.assertEqual(deleted_missing.status_code, 404)

    def test_basic_deleted(self):
        """
        When Flask has deleted paths from deleted.yaml
        check basic deleted page works
        """

        redirect = self.app_deleted.get("/deleted")

        self.assertEqual(redirect.status_code, 410)
        self.assertEqual(redirect.data, b"page deleted")

    def test_deleted_regex(self):
        """
        When Flask has deleted paths from deleted.yaml
        check deleted regex paths work
        """

        redirect = self.app_deleted.get("/deleted/nonsense/regex")

        self.assertEqual(redirect.status_code, 410)
        self.assertEqual(redirect.data, b"page deleted")

    def test_deleted_subpath_404(self):
        """
        When Flask has deleted paths from deleted.yaml
        check 404 sub-path still works
        """

        redirect = self.app_deleted.get("/deleted/missing")

        self.assertEqual(redirect.status_code, 404)

    def test_deleted_with_message(self):
        """
        Check deleted pages with extra context info
        """

        deleted = self.app_deleted.get("/deleted/with/message")

        self.assertEqual(deleted.status_code, 410)
        self.assertEqual(deleted.data, b"Gone, gone, gone")

    def test_deleted_callback(self):
        """
        Check deleted pages with a custom callback view function
        """

        deleted_callback = self.app_deleted_callback.get("/deleted")

        self.assertEqual(deleted_callback.status_code, 410)
        self.assertEqual(deleted_callback.data, b"custom callback")


if __name__ == "__main__":
    unittest.main()


tests/test_django.py
# Core
import os
import unittest

# Packages
import django
from django.conf import settings
from django.test import Client
from django.test.utils import override_settings

# Local
from canonicalwebteam.yaml_responses.django_helpers import (
    create_redirect_views,
    create_deleted_views,
)


this_dir = os.path.dirname(os.path.realpath(__file__))

# Mock Django
settings.configure(
    TEMPLATES=[
        {
            "BACKEND": "django.template.backends.django.DjangoTemplates",
            "DIRS": [f"{this_dir}/fixtures/django/templates"],
        }
    ]
)
django.setup()


class TestDjangoRedirects(unittest.TestCase):
    def test_missing_file(self):
        """
        When given a non-existent file path,
        create_redirect_views should return an empty list
        """

        self.assertEqual(
            create_redirect_views(path="/tmp/non-existent-file.yaml"), []
        )

    @override_settings(ROOT_URLCONF="tests.fixtures.django.redirects_urls")
    def test_not_found(self):
        """
        When Django has redirects from create_redirect_views,
        check 404s still work
        """

        django_client = Client()

        redirect_missing = django_client.get("/hello-missing")

        self.assertEqual(redirect_missing.status_code, 404)

    @override_settings(ROOT_URLCONF="tests.fixtures.django.redirects_urls")
    def test_basic_redirect(self):
        """
        When create_redirect_views is given a valid redirects.yaml file,
        and the resulting views are parsed by Django,
        check Django successfully redirects
        """

        django_client = Client()

        redirect = django_client.get("/hello")

        self.assertEqual(redirect.status_code, 302)
        self.assertEqual(redirect.get("Location"), "/world")

    @override_settings(ROOT_URLCONF="tests.fixtures.django.redirects_urls")
    def test_query_redirect(self):
        """
        When create_redirect_views is given a valid redirects.yaml file,
        and the resulting views are parsed by Django,
        check Django successfully redirects URL with query string
        """

        django_client = Client()

        redirect = django_client.get("/hello?name=world")

        self.assertEqual(redirect.status_code, 302)
        self.assertEqual(redirect.get("Location"), "/world?name=world")

    @override_settings(
        ROOT_URLCONF="tests.fixtures.django.permanent_redirects_urls"
    )
    def test_permanent_redirect(self):
        """
        Check we can customise the status code
        to return e.g. a 301 response, instead of 302
        """

        django_client = Client()

        redirect = django_client.get("/hello")

        self.assertEqual(redirect.status_code, 301)
        self.assertEqual(redirect.get("Location"), "/world")

    @override_settings(ROOT_URLCONF="tests.fixtures.django.redirects_urls")
    def test_regex_redirect(self):
        """
        When create_redirect_views is given a valid redirects.yaml file,
        and the resulting views are parsed by Django,
        check Django successfully redirects
        """

        django_client = Client()

        robin_redirect = django_client.get("/example-robin")
        peter_redirect = django_client.get("/example-peter")

        self.assertEqual(robin_redirect.status_code, 302)
        self.assertEqual(
            robin_redirect.get("Location"), "http://example.com/robin"
        )
        self.assertEqual(peter_redirect.status_code, 302)
        self.assertEqual(
            peter_redirect.get("Location"), "http://example.com/peter"
        )

    @override_settings(ROOT_URLCONF="tests.fixtures.django.redirects_urls")
    def test_homepage_view(self):
        """
        When redirects from create_redirect_views are successfully passed to
        Django, check other normal views still work.
        """

        django_client = Client()

        homepage = django_client.get("/homepage")

        self.assertEqual(homepage.content, b"hello world")


class TestDjangoDeleted(unittest.TestCase):
    def test_missing_file(self):
        """
        When given a non-existent file path,
        create_deleted_views should return an empty list
        """

        self.assertEqual(
            create_deleted_views(path="/tmp/non-existent-file.yaml"), []
        )

    def test_empty_file(self):
        """
        When given an empty file path,
        create_deleted_views should return an empty list
        """

        self.assertEqual(
            create_deleted_views(path="{this_dir}/fixtures/empty.yaml"), []
        )

    @override_settings(ROOT_URLCONF="tests.fixtures.django.deleted_urls")
    def test_basic_deleted(self):
        """
        When create_deleted_views is given a valid deleted.yaml file,
        and the resulting views are parsed by Django,
        check Django successfully shows 410 deleted page
        """

        django_client = Client()

        redirect = django_client.get("/deleted")

        self.assertEqual(redirect.status_code, 410)
        self.assertEqual(redirect.content, b"page deleted")

    @override_settings(ROOT_URLCONF="tests.fixtures.django.deleted_urls")
    def test_deleted_regex(self):
        """
        When create_deleted_views is given a valid deleted.yaml file,
        and the resulting views are parsed by Django,
        check Django successfully shows 410 deleted page for a regex path
        """

        django_client = Client()

        redirect = django_client.get("/deleted/nonsense/regex")

        self.assertEqual(redirect.status_code, 410)
        self.assertEqual(redirect.content, b"page deleted")

    @override_settings(ROOT_URLCONF="tests.fixtures.django.deleted_urls")
    def test_deleted_subpath_404(self):
        """
        When create_deleted_views is given a valid deleted.yaml file,
        and the resulting views are parsed by Django,
        check Django still shows 404 for non-matches
        """

        django_client = Client()

        redirect = django_client.get("/deleted/missing")

        self.assertEqual(redirect.status_code, 404)

    @override_settings(ROOT_URLCONF="tests.fixtures.django.deleted_urls")
    def test_deleted_with_message(self):
        """
        Check deleted pages with extra context info
        """

        django_client = Client()

        redirect = django_client.get("/deleted/with/message")

        self.assertEqual(redirect.status_code, 410)
        self.assertEqual(redirect.content, b"Gone, gone, gone")

    @override_settings(
        ROOT_URLCONF="tests.fixtures.django.deleted_callback_urls"
    )
    def test_deleted_callback(self):
        """
        Check deleted pages with a custom callback view function
        """

        django_client = Client()

        redirect = django_client.get("/deleted")

        self.assertEqual(redirect.status_code, 410)
        self.assertEqual(redirect.content, b"custom callback")


if __name__ == "__main__":
    unittest.main()


tests/fixtures/flask/app.py
# Core
import os

# Packages
from flask import Flask, render_template

# Local
from canonicalwebteam.yaml_responses.flask_helpers import (
    prepare_deleted,
    prepare_redirects,
)


def callback(context):
    return render_template("410.html", message="custom callback"), 410


this_dir = os.path.dirname(os.path.realpath(__file__))
parent_dir = os.path.dirname(this_dir)

app_redirects = Flask("redirects", template_folder=f"{this_dir}/templates")
app_permanent_redirects = Flask(
    "redirects", template_folder=f"{this_dir}/templates"
)
app_empty_redirects = Flask(
    "empty_redirects", template_folder=f"{this_dir}/templates"
)
app_deleted = Flask("deleted", template_folder=f"{this_dir}/templates")
app_empty_deleted = Flask(
    "empty_deleted", template_folder=f"{this_dir}/templates"
)
app_deleted_callback = Flask(
    "deleted_callback", template_folder=f"{this_dir}/templates"
)

app_redirects.before_request(
    prepare_redirects(path=f"{parent_dir}/redirects.yaml")
)
app_permanent_redirects.before_request(
    prepare_redirects(path=f"{parent_dir}/redirects.yaml", permanent=True)
)
app_deleted.before_request(prepare_deleted(path=f"{parent_dir}/deleted.yaml"))
app_deleted_callback.before_request(
    prepare_deleted(path=f"{parent_dir}/deleted.yaml", view_callback=callback)
)
app_empty_redirects.before_request(
    prepare_redirects(path=f"{parent_dir}/empty.yaml")
)
app_empty_deleted.before_request(
    prepare_deleted(path=f"{parent_dir}/empty.yaml")
)


@app_redirects.route("/homepage")
@app_deleted.route("/homepage")
def homepage():
    return "hello world"


canonicalwebteam/yaml_responses/flask.py
# Core
import os
import re

# External
import flask
import yaml
import yamlordereddictloader


class YamlRegexMap:
    def __init__(self, filepath):
        """
        Given the path to a YAML file of RegEx mappings like:

            hello/(?P<person>.*)?: "/say-hello?name={person}"
            google/(?P<search>.*)?: "https://google.com/?q={search}"

        Return a list of compiled Regex matches and destination strings:

            [
                (<regex>, "/say-hello?name={person}"),
                (<regex>, "https://google.com/?q={search}"),
            ]
        """

        self.matches = []

        if os.path.isfile(filepath):
            with open(filepath) as redirects_file:
                lines = yaml.load(
                    redirects_file, Loader=yamlordereddictloader.Loader
                )

                if lines:
                    for url_match, target_url in lines.items():
                        if url_match[0] != "/":
                            url_match = "/" + url_match

                        self.matches.append(
                            (re.compile(url_match), target_url)
                        )

    def get_target(self, url_path):
        for (match, target) in self.matches:
            result = match.fullmatch(url_path)

            if result:
                parts = {}
                for name, value in result.groupdict().items():
                    parts[name] = value or ""

                target_url = target.format(**parts)

                if flask.request.query_string:
                    target_url += "?" + flask.request.query_string.decode(
                        "utf-8"
                    )

                return target_url


def _deleted_callback(context):
    return flask.render_template("410.html", **context), 410


def prepare_redirects(path="redirects.yaml", permanent=False):
    """
    Create a regex map from the provided yaml file,
    and return a view function "apply_redirects" which encloses
    the maps to return a 302 redirect where relevant.

    Usage:
        import flask
        from canonicalwebteam.yaml_responses.flask import prepare_redirects
        app = flask.Flask(__name__)
        # Regular (temporary) redirects
        app.before_request(prepare_redirects())
        # Permanent redirects
        app.before_request(prepare_redirects(
            path='permanent_redirects.yaml', permanent=True
        ))
    """

    redirect_map = YamlRegexMap(path)

    def _apply_redirects():
        """
        Process the two mappings defined above
        of permanent and temporary redirects to target URLs,
        to send the appropriate redirect responses
        """

        redirect_url = redirect_map.get_target(flask.request.path)

        return_code = 301 if permanent else 302

        if redirect_url:
            return flask.redirect(redirect_url, code=return_code)

    return _apply_redirects


def prepare_deleted(path="deleted.yaml", view_callback=_deleted_callback):
    """
    Handlers to return 410 responses for deleted URLs loaded from
    deleted.yaml

    Basic usage:
        import flask
        from canonicalwebteam.yaml_responses.flask import prepare_deleted
        app = flask.Flask(__name__)
        app.before_request(prepare_deleted())  # Will load 410.html

    Custom usage:
        def deleted_callback(context):
            context['extra_arg'] = 'value'

            return render_template('410.html', **context), 410

        app.before_request(
            prepare_deleted(
                path="deleted.yaml",
                view_callback=deleted_callback
            )
        )
    """

    deleted_urls = {}

    if os.path.isfile(path):
        with open(path) as deleted_file:
            deleted_urls = yaml.load(deleted_file)

    def _show_deleted():
        """
        Process the two mappings defined above
        of permanent and temporary redirects to target URLs,
        to send the appropriate redirect responses
        """

        if deleted_urls:
            for url_match, context in deleted_urls.items():
                url_match = str(url_match)

                if url_match[0] != "/":
                    url_match = "/" + url_match

                if not context:
                    context = {}

                if re.compile(url_match).fullmatch(flask.request.path):
                    return view_callback(context)

    return _show_deleted


canonicalwebteam/yaml_responses/django.py
# Core packages
import os

# Third party packages
import yaml
from django.shortcuts import redirect, render
from django.conf.urls import url


def _create_view(view_callback, url_mapping, settings={}):
    """
    Create a view function to execute the callback function with the
    url_arguments
    """

    def url_view(request, *args, **kwargs):
        return view_callback(request, url_mapping, settings, *args, **kwargs)

    return url_view


def _create_views_from_yaml(yaml_filepath, view_callback, settings={}):
    """
    Givan a YAML file mapping URL paths to values, e.g.:

        path/one: {"some": "value"}
        path/two: {"another": "value"}

    Create a Django URL pattern from each value, so that when that path
    is requested, view_callback is run, passing the mapped value.

    Returns a list of Django urlpatterns.
    """

    urlpatterns = []

    if os.path.isfile(yaml_filepath):
        with open(yaml_filepath) as yaml_paths_file:
            url_paths = yaml.load(yaml_paths_file.read())
            if url_paths:
                for url_path, url_mapping in url_paths.items():
                    urlpatterns.append(
                        url(
                            r"^{0}$".format(url_path),
                            _create_view(view_callback, url_mapping, settings),
                        )
                    )

    return urlpatterns


def _redirect_to_target(request, url_mapping, settings, *args, **kwargs):
    location = url_mapping.format(**kwargs)
    query = request.META["QUERY_STRING"]

    if query:
        location += "?" + query

    return redirect(location, permanent=settings.get("permanent", False))


def _deleted_callback(request, url_mapping, settings, *args, **kwargs):
    return render(request, "410.html", url_mapping, status=410)


def create_redirect_views(path="redirects.yaml", permanent=False):
    return _create_views_from_yaml(
        path, _redirect_to_target, settings={"permanent": permanent}
    )


def create_deleted_views(path="deleted.yaml", view_callback=_deleted_callback):
    return _create_views_from_yaml(path, view_callback)


canonicalwebteam/yaml_responses/views_from_yaml.py
import yaml

from django.conf.urls import url


def _create_view(url_settings, view_callback):
    """
    Create a view function to execute the callback function with the
    url_settings
    """

    def url_view(request, *args, **kwargs):
        return view_callback(request, url_settings, *args, **kwargs)

    return url_view


def create_views_from_file(yaml_filepath, view_callback):
    """
    Givan a YAML file mapping URL paths to values, e.g.:

        path/one: {"some": "value"}
        path/two: {"another": "value"}

    Create a Django URL pattern from each value, so that when that path
    is requested, view_callback is run, passing the mapped value.

    Returns a list of Django urlpatterns.
    """

    urlpatterns = []

    with open(yaml_filepath) as yaml_paths_file:
        url_paths = yaml.load(yaml_paths_file.read())
        if url_paths:
            for url_path, url_settings in url_paths.items():
                urlpatterns.append(
                    url(
                        r'^{0}$'.format(url_path),
                        _create_view(url_settings, view_callback)
                    )
                )

    return urlpatterns


canonicalwebteam/yaml_responses/flask_helpers.py
# Standard library
import os
import re
from urllib.parse import urlparse

# Packages
import flask
import yaml
from yamlloader import ordereddict


class YamlRegexMap:
    def __init__(self, filepath):
        """
        Given the path to a YAML file of RegEx mappings like:

            hello/(?P<person>.*)?: "/say-hello?name={person}"
            google/(?P<search>.*)?: "https://google.com/?q={search}"

        Return a list of compiled Regex matches and destination strings:

            [
                (<regex>, "/say-hello?name={person}"),
                (<regex>, "https://google.com/?q={search}"),
            ]
        """

        self.matches = []

        if os.path.isfile(filepath):
            with open(filepath) as redirects_file:
                lines = yaml.load(redirects_file, Loader=ordereddict.CLoader)

                if lines:
                    for url_match, target_url in lines.items():
                        if url_match[0] != "/":
                            url_match = "/" + url_match

                        self.matches.append(
                            (re.compile(url_match), target_url)
                        )

    def get_target(self, url_path):
        for (match, target) in self.matches:
            result = match.fullmatch(url_path)

            if result:
                parts = {}
                for name, value in result.groupdict().items():
                    parts[name] = value or ""

                target_url = target.format(**parts)

                # Add request query parameters
                parsed_target_url = urlparse(target_url)
                target_query = parsed_target_url.query
                request_query = flask.request.query_string.decode()

                if request_query:
                    if target_query:
                        target_url = parsed_target_url._replace(
                            query=f"{target_query}&{request_query}"
                        ).geturl()
                    else:
                        target_url = parsed_target_url._replace(
                            query=request_query
                        ).geturl()

                return target_url


def _deleted_callback(context):
    return flask.render_template("410.html", **context), 410


def prepare_redirects(path="redirects.yaml", permanent=False):
    """
    Create a regex map from the provided yaml file,
    and return a view function "apply_redirects" which encloses
    the maps to return a 302 redirect where relevant.

    Usage:
        import flask
        from canonicalwebteam.yaml_responses.flask import prepare_redirects
        app = flask.Flask(__name__)
        # Regular (temporary) redirects
        app.before_request(prepare_redirects())
        # Permanent redirects
        app.before_request(prepare_redirects(
            path='permanent_redirects.yaml', permanent=True
        ))
    """

    redirect_map = YamlRegexMap(path)

    def _apply_redirects():
        """
        Process the two mappings defined above
        of permanent and temporary redirects to target URLs,
        to send the appropriate redirect responses
        """

        redirect_url = redirect_map.get_target(flask.request.path)

        return_code = 301 if permanent else 302

        if redirect_url:
            return flask.redirect(redirect_url, code=return_code)

    return _apply_redirects


def prepare_deleted(path="deleted.yaml", view_callback=_deleted_callback):
    """
    Handlers to return 410 responses for deleted URLs loaded from
    deleted.yaml

    Basic usage:
        import flask
        from canonicalwebteam.yaml_responses.flask import prepare_deleted
        app = flask.Flask(__name__)
        app.before_request(prepare_deleted())  # Will load 410.html

    Custom usage:
        def deleted_callback(context):
            context['extra_arg'] = 'value'

            return render_template('410.html', **context), 410

        app.before_request(
            prepare_deleted(
                path="deleted.yaml",
                view_callback=deleted_callback
            )
        )
    """

    deleted_urls = {}

    if os.path.isfile(path):
        with open(path) as deleted_file:
            deleted_urls = yaml.load(deleted_file, Loader=ordereddict.CLoader)

    def _show_deleted():
        """
        Process the two mappings defined above
        of permanent and temporary redirects to target URLs,
        to send the appropriate redirect responses
        """

        if deleted_urls:
            for url_match, context in deleted_urls.items():
                url_match = str(url_match)

                if url_match[0] != "/":
                    url_match = "/" + url_match

                if not context:
                    context = {}

                if re.compile(url_match).fullmatch(flask.request.path):
                    return view_callback(context)

    return _show_deleted
